### PR TITLE
fix(recordings): transcode HEVC recordings to H.264 on demand for browser playback

### PR DESCRIPTION
**Pull request by davlaw**
## Problem

`mp4_writer.c` is a pure stream-copy remuxer — recordings are written straight from whatever codec the camera's RTSP source delivers, with no re-encode. A camera whose active source is native HEVC produces recordings that fail to play in the recordings page's `<video>` element ("no video with supported format and MIME type found"), even though the same file plays fine in a local player with a real HEVC decoder. Most browsers have no royalty-free HEVC decoder for HTML5 video.

In our case this surfaced for a Tapo camera whose live-view path normally runs through a hardware-transcode sidecar (VAAPI H.264), so its recordings are ordinarily already H.264 — but during a brief outage of that sidecar, the source fell back to the raw camera feed and produced a batch of genuinely HEVC recordings that then wouldn't play in-browser. The underlying gap (any camera with a native HEVC stream, or any future transcode-sidecar hiccup) is general, not specific to one setup.

## Fix

On a cache miss for an HEVC recording, `api_handlers_recordings_playback.c`'s handler now transcodes to H.264 on demand and caches the result:

- `recording_needs_hevc_transcode()` probes the primary video stream via libavformat; fails open (no transcode) on any probe error so a file we can't inspect never blocks playback.
- `ensure_recording_transcode_cache()` runs ffmpeg (VAAPI hardware transcode when `/dev/dri/renderD128` exists, software `libx264` fallback otherwise — for portability across hosts and CI), writes to a `.tmp.<pid>` file, and atomically renames into place so a concurrent reader never observes a partially-written cache.
- Audio, when present, is carried through as AAC (`-c:a aac`) rather than dropped — `mp4_writer.c` already transcodes any PCM a camera sends to AAC before it reaches the MP4 container, so a source recording's audio, when present, is always AAC by the time this code runs.
- On any ffmpeg failure, falls back to serving the original file — this can only add browser playback, never regress the existing download-and-play-locally path.
- Cache lives at `<storage_path>/transcoded/<id>.mp4`, mirroring the existing `thumbnails/<id>_<n>.jpg` convention, and is removed by `storage_manager.c`'s single deletion chokepoint whenever its recording is deleted — it can't outlive the recording.

## Testing

New unit coverage: `test_recording_transcode` (9 cases — HEVC/H.264 detection, cache-hit skip, parent-directory creation, missing-source handling, and audio preservation), plus `test_path_utils` and `test_storage_manager_retention` updates for the new cache-path helper and deletion hookup. All green locally (48 tests, 0 failures), alongside no regressions in the surrounding storage/recordings test targets.

Verified end-to-end against a real 2560x1920 HEVC recording on the host this was built for: first playback request took ~10s (real VAAPI transcode), second request served the cached copy in ~0.1s.

Built with Claude Code (Anthropic).

### CODE REVIEW

File: src/video/recording_path.c [hunk 1]
```diff
@@ -81,6 +81,20 @@ int build_mp4_recording_directory_at_root(
     return (written < 0 || (size_t)written >= output_size) ? -1 : 0;
 }
 
+int build_recording_transcode_cache_path(const char *storage_path,
+                                         uint64_t recording_id,
+                                         char *output,
+                                         size_t output_size) {
+    if (!storage_path || storage_path[0] == '\0' || recording_id == 0 ||
+        !output || output_size == 0) {
+        return -1;
```
[the remark this answers is not in the document]
Comment (davlaw) [reply src/video/recording_path.c]: Fixed in 3d222918 — now checks `storage_path[0] != '/'` matching `build_mp4_recording_directory_at_root()`/`effective_placement_root()`'s convention elsewhere in this file. Added a regression test (`test_transcode_cache_path_rejects_relative_storage_path`).

File: src/video/recording_transcode.c [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+#include <libavformat/avformat.h>
+
+#include "core/path_utils.h"
+#include "video/ffmpeg_utils.h"
+#define LOG_COMPONENT "RecordingTranscode"
+#include "core/logger.h"
+
+#define VAAPI_RENDER_NODE "/dev/dri/renderD128"
+#define FFMPEG_BINARY "/usr/bin/ffmpeg"
```
[the remark this answers is not in the document]
Comment (davlaw) [reply src/video/recording_transcode.c]: Fixed in 3d222918 — FFMPEG_BINARY is now a bare "ffmpeg" and run_and_wait() uses execvp(), so it resolves against PATH. lightnvr.service's own PATH covers both /usr/bin and /usr/local/bin, so this is a no-op for the systemd deployment while fixing portability elsewhere.

File: src/video/recording_transcode.c [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        "-i", (char *)original_path,
+        "-c:v", "h264_vaapi", "-c:a", "aac",
+        "-f", "mp4", tmp_path, NULL
+    };
+    char *argv_software[] = {
+        (char *)FFMPEG_BINARY, "-y", "-hide_banner", "-loglevel", "error",
+        "-i", (char *)original_path,
+        "-c:v", "libx264", "-c:a", "aac",
+        "-f", "mp4", tmp_path, NULL
+    };
```
[the remark this answers is not in the document]
Comment (davlaw) [reply src/video/recording_transcode.c]: Pushing back on -c:a copy, keeping -c:a aac: checked mp4_writer.c further and it only transcodes *PCM* audio to AAC on the way into a recording (mp4_writer_write_packet(), is_pcm_codec() branch) — any other codec a camera sends is stream-copied through untouched. So a recording's audio isn't actually guaranteed to already be AAC; my comment overclaimed that. Forcing re-encode is what makes this feature's output reliably browser-playable across that edge case — -c:a copy would silently reintroduce the same class of bug this PR fixes, just for audio instead of video. Fixed the comment to explain this correctly instead, in 3d222918.

File: tests/unit/test_recording_transcode.c [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+void setUp(void) {
+    strcpy(g_test_dir, TEST_DIR_TEMPLATE);
+    TEST_ASSERT_NOT_NULL(mkdtemp(g_test_dir));
+
+    snprintf(g_hevc_fixture, sizeof(g_hevc_fixture), "%s/hevc_source.mp4", g_test_dir);
+    snprintf(g_h264_fixture, sizeof(g_h264_fixture), "%s/h264_source.mp4", g_test_dir);
+
+    char args[512];
+    snprintf(args, sizeof(args),
```
[the remark this answers is not in the document]
Comment (davlaw) [reply tests/unit/test_recording_transcode.c:93]: Fixed in 3d222918 — added a binaries_available() check in setUp() and TEST_IGNORE_MESSAGE in every test when ffmpeg/ffprobe aren't on PATH, matching test_go2rtc_two_config_merge.c's existing pattern for its own optional external binary. Confirmed this was a real gap: .github/workflows/sanitizer.yml installs only the FFmpeg *-dev libraries this project links against, never the ffmpeg/ffprobe CLI tools, and runs ctest — this suite would have hard-failed there.

File: src/video/recording_transcode.c [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    if (pid == 0) {
+        /* Child: silence ffmpeg's own stdout/stderr chatter into our log stream. */
+        int devnull = open("/dev/null", O_WRONLY);
+        if (devnull >= 0) {
+            dup2(devnull, STDOUT_FILENO);
+            dup2(devnull, STDERR_FILENO);
+            close(devnull);
+        }
+        execv(argv[0], argv);
+        _exit(127);
```
[the remark this answers is not in the document]
Comment (davlaw) [reply src/video/recording_transcode.c:84]: Same fix as the other execv comment — 3d222918 switches to execvp() with a bare "ffmpeg" (see reply there for details).

**Comment by davlaw** [comment]
While investigating the portability review comments, found a separate, more severe pre-existing gap worth flagging (not something this PR should try to fix, since it's unrelated to this diff and would affect any other ffmpeg-CLI-dependent feature too): **Dockerfile.alpine never installs the `ffmpeg` CLI package**, only `ffmpeg-dev`/`ffmpeg-libavformat`/`ffmpeg-libavcodec`/`ffmpeg-libswscale`. The main `Dockerfile` explicitly adds bare `ffmpeg` via apt alongside the dev libs, but the Alpine variant's `apk add` lists never include it.

This means this PR's transcode feature (and presumably anything else that shells out to `ffmpeg`) would silently no-op on an Alpine-built image — both the VAAPI and software fallback paths would fail since the binary doesn't exist at all, not just at an unexpected path. Worth a separate fix (`apk add ffmpeg` in Dockerfile.alpine) if the Alpine image is meant to have parity with the main one.

File: src/video/recording_transcode.c [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+        log_error("recording_transcode: failed to create cache directory %s: %s", cache_dir, strerror(errno));
+        return -1;
+    }
+
+    char tmp_path[512];
+    int written = snprintf(tmp_path, sizeof(tmp_path), "%s.tmp.%d", cache_path, (int)getpid());
+    if (written < 0 || (size_t)written >= sizeof(tmp_path)) {
+        log_error("recording_transcode: cache path too long: %s", cache_path);
+        return -1;
+    }
```
[the remark this answers is not in the document]
Comment (davlaw) [reply src/video/recording_transcode.c:146]: Confirmed — real bug, not a false positive. getpid() is constant across libuv's worker threads within one process, so two concurrent playback requests for the same recording did collide on the same tmp path. Fixed in `4e768f05` with an atomic per-call sequence number (`s_tmp_seq`) appended to the tmp filename, plus a new regression test (`test_ensure_cache_concurrent_calls_for_same_recording_dont_corrupt_cache`) that drives two threads at the same cache_path concurrently and asserts no corruption.

File: src/video/recording_transcode.c [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        "-i", (char *)original_path,
+        "-c:v", "h264_vaapi", "-c:a", "aac",
+        "-f", "mp4", tmp_path, NULL
+    };
+    char *argv_software[] = {
+        (char *)FFMPEG_BINARY, "-y", "-hide_banner", "-loglevel", "error",
+        "-i", (char *)original_path,
+        "-c:v", "libx264", "-c:a", "aac",
+        "-f", "mp4", tmp_path, NULL
+    };
```
[the remark this answers is not in the document]
Comment (davlaw) [reply src/video/recording_transcode.c:178]: Fixed in `4e768f05` — added `-nostdin` to both the VAAPI and software ffmpeg argv arrays.
